File: 04 architecture/01 stack machine.rb
#!/usr/bin/env ruby

require_relative "vm"

class Adder < VM
	def interpret
		@s = []
		super
	end

	def push
		@s.push(read_program)
		read_program.call
	end

	def add
		@s[1] += @s[0]
		@s = @s.drop(1)
		read_program.call
	end

	def jump_if_not_zero
		if @s[0] == 0
			@pc += 1
		else
			@pc = @program[@pc]
		end
		read_program.call
	end

	def exit
		throw :program_complete
	end

	def print_state
		puts "#{@pc}: @s => #{@s}"
	end
end

Adder.new(
	:push, 13,
	:push, -1,
	:add,
	:print_state,
	:read_program,
	:print_state,
	:jump_if_not_zero, 2,
	:exit
).interpret

File: 04 architecture/01 stack machine.go
package main

import "fmt"
import "os"

type S []int

type StackMachine struct {
	S
	VM
}

func (s *StackMachine) String() string {
	return fmt.Sprintf("%v, %v", s.VM, s.S)
}

func (s *StackMachine) Push() {
	s.PC++
	s.S = append(s.S, s.read_program().(int))
}

func (s *StackMachine) Add() {
	s.S[len(s.S) - 2] += s.S[len(s.S) - 1]
	s.S = s.S[:len(s.S) - 1]
}

func (s *StackMachine) JumpIfNotZero() {
	s.PC++
	if s.S[len(s.S) - 1] != 0 {
		s.PC = s.m[s.PC].(int)
	}
}

func main() {
	s := new(StackMachine)
	print_state := func() {
		fmt.Println(s)
	}
	skip := func() { s.PC++ }
	s.Load(
		s.Push, 13,
		Label("decrement"),
		func() { s.S = append(s.S, -1) },
		s.Add,
		print_state,
		skip,
		print_state,
		s.JumpIfNotZero, Label("decrement"),
		print_state,
		func() { os.Exit(0) },
	).Run()
}